### PR TITLE
Fake data

### LINKED ISSUE #77: Add faking framework
Use a faking framework to reduce the amount of hard-coded responses from the BAW server. This will reduce update times when changes are made to the BAW server and minimize unit test files. Potential framework: https://github.com/marak/faker.js

### PR DESCRIPTION
**Pull request by Allcharles**
# Generate fake data for models

Implement faker.js library so that faked data for models can be automatically generated.

## Changes

- Created `generate[INSERT MODEL NAME]` functions for each of the models. This generates the JSON data required to build a new model using randomised data
  - ie. `generateProject()`
  - Provide an id value to the function to set the id of the model. ie. `generateProject(1)`

## Issues

- Need to add changes into unit tests still (separate PR)
- Faker.js library last release was 2017, would like to use the 5.0.0 when it potentially comes out: https://github.com/Marak/faker.js/issues/692

## Closes

Closes #77

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -46,6 +46,7 @@
     "@ngx-loading-bar/http-client": "^4.2.0",
     "@swimlane/ngx-datatable": "^17.0.0",
     "bootstrap": "^4.5.0",
+    "faker": "^4.1.0",
```
Comment (atruskie) [inline package.json]: move this to devDependencies

File: src/app/models/AnalysisJob.ts [hunk 2]
```diff
@@ -85,6 +95,45 @@ export class AnalysisJob extends AbstractModel implements IAnalysisJob {
...
+        queued: 1,
+        working: 0,
+        success: 0,
+        failed: 0,
+        total: 1,
+      },
+      overallProgressModifiedAt: modelData.timestamp(),
+      overallCount: modelData.random.number(100),
+      overallDurationSeconds: modelData.random.arrayElement([15, 30, 60]),
+      overallDataLengthBytes: modelData.random.arrayElement([256, 512, 1024]),
```
Comment (atruskie) [inline src/app/models/AnalysisJob.ts]: if you're looking for more realistic numbers here, try a year (3.154e+7) and 500GB, and like a thousand items

File: src/app/models/AudioEvent.ts [hunk 3]
```diff
@@ -59,6 +60,26 @@ export class AudioEvent extends AbstractModel implements IAudioEvent {
...
   public audioRecording?: AudioRecording;
 
+  public static generate(id?: Id): IAudioEvent {
+    const hertzRange = [22050, 8000, 44100, 48000];
+
+    return {
+      id: modelData.id(id),
+      audioRecordingId: modelData.id(),
+      startTimeSeconds: modelData.seconds(),
+      endTimeSeconds: modelData.seconds(),
```
Comment (atruskie) [inline src/app/models/AudioEvent.ts]: it's a good idea to define endTime as start + seconds() so you guarantee the value is greater than start

ditto for frequency fields below

File: src/app/models/AudioRecording.ts [hunk 4]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
+  public static generate(id?: Id): IAudioRecording {
+    return {
+      id: modelData.id(id),
+      uuid: modelData.uuid(),
+      uploaderId: modelData.id(),
+      recordedDate: modelData.timestamp(),
+      siteId: modelData.id(),
+      durationSeconds: modelData.id(),
+      sampleRateHertz: modelData.id(),
+      bitRateBps: 22050, // TODO Replace with list of possibilities
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: 705600
352800
768000
512000
256000
82000
64000
128000
1411200
124000
123000
125000
353000

File: src/app/models/AudioRecording.ts [hunk 5]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
+      mediaType: "audio/mpeg", // TODO Replace with list of possibilities
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: audio/wave, audio/flac

File: src/app/models/AudioRecording.ts [hunk 6]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
+      dataLengthBytes: 3800, // TODO Replace with list of possibilities
+      fileHash: "SHA: 2346ad27d7568ba9896f1b7da6b5991251debdf2",
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: two colons, + randomise string

"SHA256::" + modelData.hexaDecimal(256 / 4)

File: src/app/models/AudioRecording.ts [hunk 7]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
   public site?: Site;
 
+  public static generate(id?: Id): IAudioRecording {
+    return {
+      id: modelData.id(id),
+      uuid: modelData.uuid(),
+      uploaderId: modelData.id(),
+      recordedDate: modelData.timestamp(),
+      siteId: modelData.id(),
+      durationSeconds: modelData.id(),
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: should be a common duration (30, 60,  1800, 3600, 7200, 21600)

File: src/app/models/AudioRecording.ts [hunk 8]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
+      sampleRateHertz: modelData.id(),
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: 8000, 22050, 44100, 48000

File: src/app/models/AudioRecording.ts [hunk 9]
```diff
@@ -82,6 +83,39 @@ export class AudioRecording extends AbstractModel implements IAudioRecording {
...
+      bitRateBps: 22050, // TODO Replace with list of possibilities
+      mediaType: "audio/mpeg", // TODO Replace with list of possibilities
+      dataLengthBytes: 3800, // TODO Replace with list of possibilities
```
Comment (atruskie) [inline src/app/models/AudioRecording.ts]: bit_rate_bps * duration_seconds

File: src/app/models/DatasetItem.ts [hunk 10]
```diff
@@ -45,6 +46,19 @@ export class DatasetItem extends AbstractModel implements IDatasetItem {
...
 
+  public static generate(id?: Id): IDatasetItem {
+    return {
+      id: modelData.id(id),
+      datasetId: modelData.id(),
+      audioRecordingId: modelData.id(),
+      creatorId: modelData.id(),
+      createdAt: modelData.timestamp(),
+      startTimeSeconds: modelData.seconds(),
+      endTimeSeconds: modelData.seconds(),
```
Comment (atruskie) [inline src/app/models/DatasetItem.ts]: might be worth making the start/end generation a helper

File: src/app/models/ProgressEvent.ts [hunk 11]
```diff
@@ -33,6 +34,16 @@ export class ProgressEvent extends AbstractModel implements IProgressEvent {
   @HasOne<ProgressEvent>(DATASET_ITEM, "datasetItemId")
   public datasetItem?: DatasetItem;
 
+  public static generate(id?: Id): IProgressEvent {
+    return {
+      id: modelData.id(id),
+      creatorId: modelData.id(),
+      datasetItemId: modelData.id(),
+      activity: "viewed", // TODO Replace with list of possibilities
```
Comment (atruskie) [inline src/app/models/ProgressEvent.ts]: viewed, played, annotated

File: src/app/models/Question.ts [hunk 12]
```diff
@@ -36,6 +37,18 @@ export class Question extends AbstractModel implements IQuestion {
   @Updater<Question>()
   public updater?: User;
 
+  public static generate(id?: Id): IQuestion {
+    return {
+      id: modelData.id(id),
+      text: modelData.notes(), // TODO This may not be the correct type
+      data: modelData.notes(), // TODO This may not be the correct type
```
Comment (atruskie) [inline src/app/models/Question.ts]: data will be an arbitrary object (correct)

text will be a html/textual blurb (incorrect)

File: src/app/models/Script.ts [hunk 13]
```diff
@@ -50,13 +51,36 @@ export class Script extends AbstractModel implements IScript {
...
   @Creator<Script>()
   public creator?: User;
 
+  public static generate(id?: Id): IScript {
+    return {
+      id: modelData.id(id),
+      name: modelData.param(),
+      description: modelData.description(),
+      analysisIdentifier: "script machine identifier", // TODO Implement with random values
+      version: modelData.random.number(50) / 10,
```
Comment (atruskie) [inline src/app/models/Script.ts]: use faker.system.semver

File: src/app/models/TagGroup.ts [hunk 14]
```diff
@@ -40,6 +41,16 @@ export class TagGroup extends AbstractModel implements ITagGroup {
   @HasOne<TagGroup>(TAG, "tagId")
   public tag?: Tag;
 
+  public static generate(id?: Id): ITagGroup {
+    return {
+      id: modelData.id(id),
+      groupIdentifier: modelData.random.uuid(), // TODO Check
```
Comment (atruskie) [inline src/app/models/TagGroup.ts]: just any common name, e.g. "boobook owl"

File: src/app/test/helpers/faker.ts [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+  notes: () => randomObject(0, 5),
+  offset: () =>
+    faker.random.arrayElement(["+", "-"]) +
+    faker.random.number(11) +
+    ":" +
+    faker.random.arrayElement(["00", "30"]),
+  param: () => faker.name.title(),
+  seconds: () => faker.random.number(86400 - 30) + 30,
+  timestamp: () => faker.date.past().toISOString(),
+  timezone: () => undefined, // TODO
```
Comment (atruskie) [inline src/app/test/helpers/faker.ts]: depending on where this is used it'll be something like `australia/sydney'

**Review by atruskie — CHANGES_REQUESTED** [review]
OK, the big change that will be mandatory here is that the faker dependency most not be depended on in production code. We don't want faker, or even the methods for generating fake data, in any app code. They are test only concerns.

The rest of the comments are just about making the fakes more realistic

File: src/app/test/helpers/faker.ts (hunk 15 above)
Comment (Allcharles) [reply src/app/test/helpers/faker.ts]: My goal is to use timezones which are different to the one the machine is currently in. That way i can properly test conversions are done.

File: src/app/test/fakes/AnalysisJob.ts [hunk 16]
```diff
@@ -0,0 +1,46 @@
+import { Id } from "@interfaces/apiInterfaces";
+import { AnalysisJobStatus, IAnalysisJob } from "@models/AnalysisJob";
+import { modelData } from "@test/helpers/faker";
+
+export function generateAnalysisJob(id?: Id): IAnalysisJob {
+  const overallDurationSeconds = modelData.random.number(Number("3.154E7")); // 1 year
```
Comment (atruskie) [inline src/app/test/fakes/AnalysisJob.ts]:
```suggestion
  const overallDurationSeconds = modelData.random.number(3.154e7); // 1 year
```

File: src/app/test/fakes/AnalysisJob.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  const overallDataLengthBytes = overallDurationSeconds * 22050; // duration seconds * bit rate per second
```
Comment (atruskie) [inline src/app/test/fakes/AnalysisJob.ts]:
```suggestion
  const overallDataLengthBytes = overallDurationSeconds * 22050 * 2; // duration seconds * sample_rate * two bytes per sample
```

File: src/app/test/helpers/faker.ts [hunk 18]
```diff
@@ -0,0 +1,192 @@
...
+      "c",
+      "d",
+      "e",
+      "f",
+      "A",
+      "B",
+      "C",
+      "D",
+      "E",
+      "F",
```
Comment (atruskie) [inline src/app/test/helpers/faker.ts]: no reason to have mixed case here. we usually default to lower case

**Review by atruskie — APPROVED** [review]
Minor changes suggested, otherwise approved

File: src/app/test/helpers/faker.ts (hunk 18 above)
Comment (Allcharles) [reply src/app/test/helpers/faker.ts]: Twas just a copy paste from the faker.js implementation, if they ever make a new release it'll change back to this. I'll make the change anyways.

**Pull request merged**